ctapipe/tools/muon_reconstruction.py

```diff
@@ -5,121 +5,177 @@
+        CameraCalibrator, EventSource, MuonRingFitter, MuonIntensityFitter,
+    ]
 
-    aliases = t.Dict({
+    aliases = {
         'input': 'EventSource.input_url',
-        'outfile': 'MuonDisplayerTool.outfile',
-        'display': 'MuonDisplayerTool.display',
+        'outfile': 'MuonAnalysis.outfile',
         'max_events': 'EventSource.max_events',
         'allowed_tels': 'EventSource.allowed_tels',
-    })
+    }
 
     def setup(self):
-        self.source: EventSource = self.add_component(
-            EventSource.from_config(parent=self)
-        )
-        if self.source.input_url == '':
-            raise ToolConfigurationError("please specify --input <events file>")
+        self.source = self.add_component(EventSource.from_config(parent=self))
         self.calib = self.add_component(CameraCalibrator(
             parent=self,
             subarray=self.source.subarray
         ))
-        self.writer = self.add_component(HDF5TableWriter(self.outfile, "muons"))
+        self.ring_fitter = self.add_component(MuonRingFitter(
+            parent=self,
+        ))
+        self.intensity_fitter = self.add_component(MuonIntensityFitter(
+            parent=self,
+            subarray=self.source.subarray,
+        ))
+        self.cleaning = self.add_component(
+            TailcutsImageCleaner(
+                parent=self,
+                subarray=self.source.subarray,
+            )
+        )
+        if self.outfile:
+            self.writer = self.add_component(HDF5TableWriter(
+                self.outfile, "dl1", add_prefix=True
+            ))
+        self.pixels_in_tel_frame = {}
+        self.min_pixels.attach_subarray(self.source.subarray)
 
     def start(self):
+        for event in tqdm(self.source, desc='Processing events: '):
+            self.process_array_event(event)
+
+    def process_array_event(self, event):
+        self.calib(event)
+        for tel_id, dl1 in event.dl1.tel.items():
+            self.process_telescope_event(event.index, tel_id, dl1)
+
+        if self.outfile:
+            self.writer.write('mc', [event.index, event.mc])
+
+    def process_telescope_event(self, event_index, tel_id, dl1):
+        event_id = event_index.event_id
+
+        self.log.debug(f'Processing event {event_id}, telescope {tel_id}')
+        image = dl1.image
+        clean_mask = self.cleaning(tel_id, image)
+
+        if np.count_nonzero(clean_mask) <= self.min_pixels.tel[tel_id]:
+            self.log.debug(
+                f'Skipping event {event_id}-{tel_id}:'
+                f' has less then {self.min_pixels.tel[tel_id]} pixels after cleaning'
+            )
+            return
+
+        if tel_id not in self.pixels_in_tel_frame:
+            self.pixels_in_tel_frame[tel_id] = self.pixel_to_telescope_frame(tel_id)
+
+        pixel_coords = self.pixels_in_tel_frame[tel_id]
+        x = pixel_coords.delta_az
+        y = pixel_coords.delta_alt
+
+        mask = clean_mask
+        for i in range(3):
```

Comment:
Is there a reason this is repeated 3 times? If so, please add a comment to explain it (otherwise it will be removed).

Reply:
What do you mean it will be removed?

It's an iterative fit procedure (the mask is updated with the result from the fit before)

Reply:
but yes, I can add a comment.

Reply:
I mean, if I say that,  would assume it's a mistake and remove it. Just adding a comment that it runs the fit multiple times to iterate is enough, I think.  It's not obvious that the loop does something different each time.

Reply:
Might also comment why 3 times is a good enough number of iterations

Reply:
I did not evaluate this. This comes from allison's and ruben's codes. But looking at the iterations for some muon events, it does not change a lot between 2 and three. Better would be some kind of convergence criterium with max number of iterations. But that's is also something that could be addressed in another PR after this has a reasoable structure.

This PR is more about transforming the mess in `muon_reco_functions` into a nice tool, not improving the algorithms themselves.